### PR TITLE
Initial classes

### PR DESCRIPTION
**Pull request by mecha**
Adds the initial classes and traits. Actually just traits.

### CODE REVIEW

File: src/ContainerGetCapableTrait.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            return $container[$key];
+        }
+
+        if ($isObject && property_exists($container, $key)) {
+            return $container->{$key};
+        }
+
+        if ($isTraversable) {
+            try {
+                $iterator = $this->_resolveIterator($container);
```
Comment (XedinUnknown) [inline src/ContainerGetCapableTrait.php]: Not sure if this is actually necessary. You are not requiring the specific `Iterator` API here, nor are you looking for something custom. You are just iterating, and for that `Traversable` is already enough.

File: src/ContainerGetCapableTrait.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                    if ($_key === $key) {
+                        return $_value;
+                    }
+                }
+            } catch (OutOfRangeException $outOfRangeException) {
+                // Do nothing. The iterator was not resolved, so the key was not found.
+            }
+        }
+
+        throw $this->_createNotFoundException($this->__('Key "%s" was not found', [$key]), null, null, null, $key);
```
Comment (XedinUnknown) [inline src/ContainerGetCapableTrait.php]: No need to put the key into the description. It's already part of the exception's data.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
Overall, very good. Note the two comments.

Also, I'm not sure whether a `Traversable` should qualify as a container. IMHO, when something wants to be seen as a container, it must be an array/object, or explicitly implement `ContainerInterface` or `ArrayAccess`. A `Traversable` is meant to represent a list, i.e. something to be iterated over. When something wants to be both, it can - such as the `Map` that I am building right now.

**Comment by mecha** [comment]
@XedinUnknown Ideally, yes. But what do you do when an implementation does **not** implement `ArrayAccess`, `ContainerInterface` or the like? You say that it should if it is to be considered as a container.

Let's say an object provided by a third part _**is**_ a container, just not one with a known API, what do you do? Must you create a wrapping implementation? If so, why doesn't the argument extend to wrapping arrays and objects and we just expect `ContainerInterface`? That would of course defeat the purpose of this package.

Of course, you can't know _every_ possible API. So the aim of this package is to use all the known ones. Hence my view is this: if I can get a value that corresponds to a specific key from an object, that object deserves to be considered a container. I think `Traversable` is a very good fallback. It makes the functionality compatible with a large set of known APIs.

FYI. `Traversable` is not meant to represent a "list" - it's just an interface for something that can be traversed using PHP's `foreach`. This is not my personal opinion. [This is PHP's documentation](http://php.net/manual/en/class.traversable.php).

File: src/ContainerGetCapableTrait.php (hunk 2 above)
Comment (mecha) [reply src/ContainerGetCapableTrait.php]: True, but it adds meaning to the exception message, which is _meant_ to be meaningful.

File: src/ContainerGetCapableTrait.php (hunk 1 above)
Comment (mecha) [reply src/ContainerGetCapableTrait.php]: You're right. I'll make the changes.

**Comment by XedinUnknown** [comment]
> But what do you do when an implementation does not implement `ArrayAccess`, `ContainerInterface` or the like? You say that it should if it is to be considered as a container.

Yep, that's what I'm saying.

> ... why doesn't the argument extend to wrapping arrays and objects and we just expect ContainerInterface?

Because there are other things that are considered to be a container, not just `ContainerInterface`.

> I think `Traversable` is a very good fallback. It makes the functionality compatible with a large set of known APIs.

I don't see how. What APIs do you know, other than the ones we've accounted for, that provide an `iterable` interface, but do not expose a more specific API for retrieving values by key?

> FYI. `Traversable` is not meant to represent a "list" - it's just an interface for something that can be traversed using PHP's `foreach`.

Something that can be traversed using PHP's `foreach` is a list. Don't care how many things there are (0 or more), don't care what specific keys there are (each handled separately); only that there are some non-negative number of things, and that they come in some order.

But why am I really against using iterables as containers? Because it will encourage bad code. We and other users may get tempted to pass iterables as containers. Many of them will check if something `has()` the element first, then retrieve it. This leads to double the iteration. On a large list, this can take a lot of resources. So, if something wants to be a container, I would prefer that it is explicit. Truly, for most things this is quite simple: cast to an array, enumerate the properties, wrap in a `ContainerInterface`; the options are many. But this will guarantee performant and structured retrieval of values.

**Comment by mecha** [comment]
> Something that can be traversed using PHP's foreach is a list.

Something that can be traversed is a `Traversable`. The _items yielded_ from traversing form a "list". I can write a `class FooMentalDogeGenerator implements Iterator, HuggableInterface` that can `hug()` and generates 10 random numbers when iterated. An object instance of this class is not a list. What I get when I iterate over it is a list.

> What APIs do you know, other than the ones we've accounted for, that provide an iterable interface, but do not expose a more specific API for retrieving values by key?

Confusing question, since almost everything will provide a specific API. The problem is that we cannot account for every API. If you want real-life examples, I have none to provide. But a hypothetical one that comes to mind is a PSR-6 compliant caching implementation that is iterable.

> But why am I really against using iterables as containers? Because it will encourage bad code. We and other users may get tempted to pass iterables as containers. Many of them will check if something has() the element first, then retrieve it. This leads to double the iteration.

Temptation is not an argument against, IMO.

But you're right, that's nonsensical. And also the responsibility of the consumer of these traits. Independently, these traits don't care. The `ContainerGetCapableTrait` is not concerned whether the `use`r will be calling `ContainerHasCapableTrait::_containerHas()`.

A good idea is to explicitly document and mention that `Traversable` arguments are iterated, and thus performance should be kept in mind when using them.

> So, if something wants to be a container, I would prefer that it is explicit.

Me too. But this is a different argument.

These traits allow the classes that use them to read from anything that has a key that corresponds to a value. No matter what the container is. These traits should not define what a container is or should be. The using class is the one that knows what it is dealing with.

Be it something as simple as an array or some iterable Symfony object , the aim here is to provide functionality that can retrieve/check a value from something that can associate a key to a value in a vanilla PHP manner.

File: src/ContainerGetCapableTrait.php (hunk 2 above)
Comment (XedinUnknown) [reply src/ContainerGetCapableTrait.php]: Without this, the description already says what is wrong. **With** what - that's the job of [`DataKeyAwareInterface#getDataKey()`].

[`DataKeyAwareInterface#getDataKey()`]:  https://github.com/Dhii/data-container-interface/blob/develop/src/DataKeyAwareInterface.php#L21

**Comment by XedinUnknown** [comment]
> These traits allow the classes that use them to read from anything that has a key that corresponds to a value

Yes. But a `Traversable` is something that is meant to be iterated over, not something meant to provide a value for a key. That's the whole difference. If it wants to provide that, it should be a known implementation - that's the whole point of the interop Container standard, and that's why we also cater for arrays and objects. The fact that you can get a value by key from something that is not meant to do that is confusing and ambiguous. I strongly feel that it should not be.

**Comment by mecha** [comment]
I may not fully agree, but fine I'll remove the `Traversable` features.

What are your thoughts on checking if objects have `__get()` and `__isset()` magic functionality? PHP seems to handle such objects differently from regular `stdClass` instances. So they require special handling.

**Comment by XedinUnknown** [comment]
Thanks.

I think better not go into black magic. Light is our only hope.

**Comment by mecha** [comment]
haha

I'm serious though. It's a valid container, and at first I thought it would just work. But `property_exists` doesn't cater for `__isset()`. I don't think it's _thaaat_ evil to just check for them :laughing:

**Comment by XedinUnknown** [comment]
The list of valid containers is currently as follows:
- `ContainerInterface`
- `array`
- `ArrayAccess`
- `stdClass`

Catering for magic will de-stabilize the strict yet fair balance of the currents of Aether currently flowing through the code, and will make use unable to tell which objects are containers, and which are not. Telling good from evil is the blessing that Adam gave us by eating from the apple, and while we are all still being punished for his deed, it would be unwise to neglect this ability.

**Comment by mecha** [comment]
How are `object` and `stdClass` different in this context?

**Comment by XedinUnknown** [comment]
An `stdClass` is an explicit container. An `object` is any instance of anything.

**Comment by mecha** [comment]
And how can "an instance of anything" be used as a container?

**Comment by XedinUnknown** [comment]
It shouldn't. That's the point.

**Comment by mecha** [comment]
My bad. I refreshed and saw that you removed it from the list in your [previous comment](https://github.com/Dhii/container-helper-base/pull/1#issuecomment-361868179).

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
This specifically allows for not only `stdClass`, but also for magic getters, i.e. what I was against.

**Comment by mecha** [comment]
@XedinUnknown Where?

**Comment by XedinUnknown** [comment]
@mecha, [here](https://github.com/Dhii/container-helper-base/pull/1/commits/592389624ec5f595241aad935ed5a4f5aeabaf35#diff-cab753c086eb9f5cd444885be79cce75R66)

**Comment by mecha** [comment]
Oh crap. I thought that would be safeguard against `property_exists()`'s false negative. But I misread the manual:

> As opposed with isset(), property_exists() returns TRUE even if the property has the value NULL.

Will change.

**Comment by mecha** [comment]
That is of course, if `null` is a valid return. Which I assume we can both agree that it is.

**Comment by XedinUnknown** [comment]
Absolutely.

**Comment by XedinUnknown** [comment]
This is one of the benefits brought to us by `NotFoundExceptionInterface` =)

**Comment by XedinUnknown** [comment]
@mecha, to standardize on containers even tighter, we should check specifically for `stdClass`. This way, containers will be closer to what `AbstractBaseMap` works with, and will pass normalization by [`NormalizeIterableCapableTrait#_normalizeIterable()`][NormalizeIterableCapableTrait#_normalizeIterable()]. With the new approach to factories, it will also be something that is [accepted](https://github.com/Dhii/factory-interface/issues/5) by `FactoryInterface#make()`.

[NormalizeIterableCapableTrait#_normalizeIterable()]: https://github.com/Dhii/normalization-helper-base/blob/develop/src/NormalizeIterableCapableTrait.php#L30

**Review by XedinUnknown — APPROVED** [review]

**Pull request merged**